src/Data/Term.hs

```diff
@@ -1,44 +1,66 @@
 module Data.Term where
 
 import Data.Binding
+import Data.Typing
 import Data.Foldable
 import Data.Name
 import qualified Data.Set as Set
 
-data Term f = Term { freeVariables :: Set.Set Name, out :: Binding f (Term f) }
+type Result a = Either String a
+
+data Term f = Term { freeVariables :: Set.Set Name, typeOf :: Result (Term f), out :: Typing (Binding f) (Term f) }
 
 variable :: Name -> Term f
-variable name = Term (Set.singleton name) (Variable name)
+variable name = Term (Set.singleton name) (Right implicit) (Binding (Variable name))
 
 abstraction :: Name -> Term f -> Term f
-abstraction name body = Term (Set.delete name $ freeVariables body) (Abstraction name body)
+abstraction name scope = Term (Set.delete name $ freeVariables scope) (Right implicit) (Binding (Abstraction name scope))
+
+annotation :: Foldable f => Term f -> Term f -> Term f
+annotation term type' = Term (foldMap freeVariables (out term) `mappend` foldMap freeVariables (out type')) (Right type') $ Annotation term type'
+
+typing :: Foldable f => Typing (Binding f) (Term f) -> Term f
+typing t = Term (foldMap freeVariables t) (Right implicit) t
+
+binding :: Foldable f => Binding f (Term f) -> Term f
+binding b = Term (foldMap freeVariables b) (Right implicit) (Binding b)
 
 expression :: Foldable f => f (Term f) -> Term f
-expression e = Term (foldMap freeVariables e) (Expression e)
+expression e = binding $ Expression e
+
+_type :: Foldable f => Int -> Term f
+_type n = Term mempty (Right . _type $ n + 1) $ Type n
+
+_type' :: Foldable f => Term f
+_type' = _type 0
+
+implicit :: Term f
+implicit = Term mempty (Right implicit) Implicit
```

Comment:
This recursion on the type of `implicit` leaves me feeling very exposed. Laziness can be convenient but scary.